File: ios/Cloudix/GoBridge.swift
import Foundation
import Cloudixmobile
import WebKit

/// GoBridge is the whole contract between the Go core and the WebView.
///
/// Go calls in through MobileCallback (events, logs, saving a file); the page
/// calls out through a WKScriptMessageHandler. Both directions carry JSON, so
/// the existing frontend runs unchanged — see frontend/public/mobile-bridge.js,
/// which installs the window.go / window.runtime stand-ins this talks to.
final class GoBridge: NSObject {
    private weak var webView: WKWebView?
    /// The share sheet needs a view controller to present from.
    private weak var presenter: UIViewController?

    init(webView: WKWebView, presenter: UIViewController) {
        self.webView = webView
        self.presenter = presenter
    }

    /// Where the SQLite database lives. os.UserConfigDir() means nothing inside
    /// the sandbox, so the Go side is told explicitly.
    static var dataDirectory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("Cloudix", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    /// What this build can actually do.
    ///
    /// Everything switched off here is off because iOS or the signing account
    /// will not allow it, not because it is unfinished:
    ///   • screenShareSend    — ReplayKit needs a broadcast extension and an App
    ///                          Group, and App Groups need a paid account.
    ///   • lanDiscovery       — multicast needs an Apple-granted entitlement.
    ///   • networkHosting     — iOS kills a listening socket on suspend.
    ///   • backgroundDelivery — no push without a paid account and a sender.
    ///   • notifications      — same reason.
    /// The UI hides each of these rather than showing a control that fails.
    static let features: [String: Bool] = [
        "calls": true,
        "screenShareSend": false,
        "screenShareReceive": true,
        "lanDiscovery": false,
        "manualPeers": true,
        "networkHosting": false,
        "backgroundDelivery": false,
        "openDataFolder": false,
        "notifications": false,
    ]

    static var featuresJSON: String {
        guard let data = try? JSONSerialization.data(withJSONObject: features),
              let json = String(data: data, encoding: .utf8) else { return "{}" }
        return json
    }

    // MARK: page -> Go

    /// Handles one {type:"call"|"openURL"} message from the page.
    func handle(_ body: [String: Any]) {
        switch body["type"] as? String {
        case "call":
            guard let id = body["id"] as? Int,
                  let method = body["method"] as? String else { return }
            let args = body["args"] as? String ?? "[]"
            // Go can block here — a send waits on the network — so never run it
            // on the thread that is also driving the UI.
            DispatchQueue.global(qos: .userInitiated).async { [weak self] in
                var err: NSError?
                let result = MobileCall(method, args, &err)
                if let err = err {
                    self?.resolve(id: id, ok: false, payload: err.localizedDescription)
                } else {
                    self?.resolve(id: id, ok: true, payload: result)
                }
            }

        case "openURL":
            guard let raw = body["url"] as? String, let url = URL(string: raw) else { return }
            DispatchQueue.main.async { UIApplication.shared.open(url) }

        default:
            break
        }
    }

    private func resolve(id: Int, ok: Bool, payload: String) {
        let js = "window.__cloudix.resolve(\(id), \(ok), \(Self.quote(payload)))"
        evaluate(js)
    }

    private func evaluate(_ js: String) {
        DispatchQueue.main.async { [weak self] in
            self?.webView?.evaluateJavaScript(js, completionHandler: nil)
        }
    }

    /// JSON-encodes a string so it can be pasted into a JS expression. Message
    /// bodies contain quotes, newlines and base64 payloads; naive escaping
    /// breaks on all three.
    static func quote(_ s: String) -> String {
        guard let data = try? JSONEncoder().encode(s),
              let out = String(data: data, encoding: .utf8) else { return "\"\"" }
        return out
    }
}

// MARK: - Go -> page

extension GoBridge: MobileCallbackProtocol {
    func onEvent(_ name: String?, payloadJSON: String?) {
        guard let name = name else { return }
        evaluate("window.__cloudix.event(\(Self.quote(name)), \(Self.quote(payloadJSON ?? "")))")
    }

    func onLog(_ level: String?, message: String?) {
        NSLog("[cloudix/%@] %@", level ?? "info", message ?? "")
    }

    /// Writes the bytes into Documents, which UIFileSharingEnabled exposes in
    /// the Files app, and then offers a share sheet. Returning the path keeps
    /// the Go side's contract; the sheet is a convenience on top.
    func saveFile(_ suggestedName: String?, data: Data?, error: NSErrorPointer) -> String {
        guard let data = data else { return "" }
        let name = (suggestedName?.isEmpty == false ? suggestedName! : "cloudix-file")
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let url = docs.appendingPathComponent(name)
        do {
            try data.write(to: url, options: .atomic)
        } catch let failure as NSError {
            error?.pointee = failure
            return ""
        }

        DispatchQueue.main.async { [weak self] in
            guard let presenter = self?.presenter else { return }
            let sheet = UIActivityViewController(activityItems: [url], applicationActivities: nil)
            sheet.popoverPresentationController?.sourceView = presenter.view
            presenter.present(sheet, animated: true)
        }
        return url.path
    }
}

File: android/app/src/main/java/com/cloudix/messenger/GoBridge.kt
package com.cloudix.messenger

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebView
import mobile.Callback
import org.json.JSONObject
import java.io.File
import java.util.concurrent.Executors

/**
 * GoBridge is the whole contract between the Go core and the WebView.
 *
 * Go calls in through mobile.Callback (events, logs, saving a file); the page
 * calls out through the CloudixNative JavascriptInterface. Both directions
 * carry JSON, so the existing frontend runs unchanged — see
 * frontend/public/mobile-bridge.js, which installs the window.go and
 * window.runtime stand-ins this talks to.
 */
class GoBridge(private val activity: Activity, private val webView: WebView) : Callback {

    // Go can block here — a send waits on the network — so calls never run on
    // the thread that is also driving the UI.
    private val workers = Executors.newCachedThreadPool()

    companion object {
        private const val TAG = "cloudix"

        /**
         * What this build can actually do.
         *
         * Android keeps almost everything the desktop has: a foreground service
         * holds the socket open, so messages and calls arrive while the app is
         * in the background, and MediaProjection can share the screen.
         * Hosting is off because a phone behind carrier NAT can never accept an
         * inbound connection — the UI hides it rather than offering a control
         * that always fails.
         */
        fun featuresJson(): String = JSONObject(
            mapOf(
                "calls" to true,
                "screenShareSend" to true,
                "screenShareReceive" to true,
                "lanDiscovery" to true,
                "manualPeers" to true,
                "networkHosting" to false,
                "backgroundDelivery" to true,
                "openDataFolder" to false,
                "notifications" to true,
            )
        ).toString()
    }

    // MARK: page -> Go

    @JavascriptInterface
    fun post(raw: String) {
        val msg = try {
            JSONObject(raw)
        } catch (e: Exception) {
            Log.w(TAG, "bad bridge message: $raw", e)
            return
        }

        when (msg.optString("type")) {
            "call" -> {
                val id = msg.optInt("id", -1)
                val method = msg.optString("method")
                val args = msg.optString("args", "[]")
                if (id < 0 || method.isEmpty()) return
                workers.execute {
                    try {
                        resolve(id, true, mobile.Mobile.call(method, args) ?: "")
                    } catch (e: Exception) {
                        resolve(id, false, e.message ?: "call failed")
                    }
                }
            }

            "openURL" -> {
                val url = msg.optString("url")
                if (url.isEmpty()) return
                activity.runOnUiThread {
                    try {
                        activity.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
                    } catch (e: Exception) {
                        Log.w(TAG, "cannot open $url", e)
                    }
                }
            }
        }
    }

    private fun resolve(id: Int, ok: Boolean, payload: String) {
        evaluate("window.__cloudix.resolve($id, $ok, ${quote(payload)})")
    }

    private fun evaluate(js: String) {
        webView.post { webView.evaluateJavascript(js, null) }
    }

    /**
     * JSON-encodes a string so it can be pasted into a JS expression. Message
     * bodies carry quotes, newlines and base64 payloads; naive escaping breaks
     * on all three.
     */
    private fun quote(s: String): String = JSONObject.quote(s)

    // MARK: Go -> page

    override fun onEvent(name: String?, payloadJSON: String?) {
        if (name == null) return
        evaluate("window.__cloudix.event(${quote(name)}, ${quote(payloadJSON ?: "")})")
    }

    override fun onLog(level: String?, message: String?) {
        when (level) {
            "error" -> Log.e(TAG, message ?: "")
            "warn" -> Log.w(TAG, message ?: "")
            else -> Log.i(TAG, message ?: "")
        }
    }

    /**
     * Writes into the app's own Downloads directory, which is visible in the
     * Files app and needs no storage permission on any supported API level.
     */
    override fun saveFile(suggestedName: String?, data: ByteArray?): String {
        if (data == null) return ""
        val name = if (suggestedName.isNullOrEmpty()) "cloudix-file" else suggestedName
        val dir = activity.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)
            ?: activity.filesDir
        dir.mkdirs()
        val file = File(dir, name)
        file.writeBytes(data)
        return file.absolutePath
    }

    fun shutdown() {
        workers.shutdownNow()
    }
}
